### PR TITLE
Map OTEL Resources to Jaeger Process Tags

### PR DESCRIPTION
**Pull request by Grunet**
Aims to implement [this section](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk_exporters/jaeger.md#resource) of the Jaeger spec

I drew lots of inspiration from the [Go implementation](https://github.com/open-telemetry/opentelemetry-go/blob/86ced11f2c060e479817477f614438faa799ce93/exporters/jaeger/jaeger.go#L326)

I would like to try and get to some functional testing around Thrift things (since this should be the last Thrift PR for spec compliance) but I think PR review can probably happen in parallel to that

### CODE REVIEW

**Comment by tidal** [comment]
Thaynk you @Grunet !
I will take a look.

File: src/Contrib/Jaeger/HttpSender.php [hunk 1]
```diff
@@ -22,75 +26,105 @@ class HttpSender
...
+            )
+        );
+
+        $this->batchAdapterFactory = $batchAdapterFactory ?? new BatchAdapterFactory();
+    }
+
+    public function send(iterable $spans): void
+    {
+        $spansGroupedByResource = self::groupSpansByResource($spans);
+        $batchArray = $this->createBatchesPerResource($spansGroupedByResource);
```
Comment (tidal) [inline src/Contrib/Jaeger/HttpSender.php:56]: Can you please make this one call and use "batches"?
```php
 $batches = $this->createBatchesPerResource(
    self::groupSpansByResource($spans)
 );
```

File: src/Contrib/Jaeger/HttpSender.php [hunk 2]
```diff
@@ -22,75 +26,105 @@ class HttpSender
...
+
+            $process = $this->createProcessFromResource($resource);
+            $convertedSpans = (new SpanConverter())->convert($spans);
+
+            $batch = $this->batchAdapterFactory->createBatchAdapter([
+                'spans' => $convertedSpans,
+                'process' => $process,
+            ]);
+
+            $batches[] = $batch;
```
Comment (tidal) [inline src/Contrib/Jaeger/HttpSender.php:99]: Could you make this one call, please?
```php
$batches[] =  $this->batchAdapterFactory->createBatchAdapter([
    'spans' => (new SpanConverter())->convert(
        $dataForBatch['spans']
    ),
    'process' => $this->createProcessFromResource(
        $dataForBatch['resource']
    ),
]);
```

File: src/Contrib/Jaeger/TagFactory/TagFactory.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Jaeger\TagFactory;
+
+use Jaeger\Thrift\Tag;
+use Jaeger\Thrift\TagType;
+
+class TagFactory
+{
+    public static function createTag(string $key, $value): Tag
```
Comment (tidal) [inline src/Contrib/Jaeger/TagFactory/TagFactory.php:12]: See above (factory method name)

File: src/SDK/Resource/ResourceInfo.php [hunk 4]
```diff
@@ -38,4 +38,19 @@ public function getSchemaUrl(): ?string
...
+        ksort($copyOfAttributesAsArray); //sort the associative array by keys since the serializer will consider equal arrays different otherwise
+
+        $dehydratedAsArray = [
+            'schemaUrl' => $this->schemaUrl,
+            'attributes' => $copyOfAttributesAsArray,
+        ];
+
+        $serializedAsString = serialize($dehydratedAsArray);
+
+        return $serializedAsString; //The exact value doesn't matter, as long as it can distingusih between instances that represent the same/different resources
```
Comment (tidal) [inline src/SDK/Resource/ResourceInfo.php:54]: Can you please make this one call?
```php
return serialize([
    'schemaUrl' => $this->schemaUrl,
    'attributes' => $copyOfAttributesAsArray,
]);
```

File: tests/Unit/Contrib/JaegerHttpSenderTest.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        $this->assertSame(0, count($interceptedVals[0]['process']->tags));
+
+        //2nd batch
+        $this->assertSame(1, count($interceptedVals[1]['spans'])); //Detailed tests for the conversion live elsewhere
+
+        $this->assertSame('nameOfThe2ndLogicalApp', $interceptedVals[1]['process']->serviceName);
+
+        $this->assertSame(1, count($interceptedVals[1]['process']->tags));
+        $this->assertSame('telemetry.sdk.name', $interceptedVals[1]['process']->tags[0]->key);
+        $this->assertSame('opentelemetry', $interceptedVals[1]['process']->tags[0]->vStr);
```
Comment (tidal) [inline tests/Unit/Contrib/JaegerHttpSenderTest.php:124]: Could you please split these asserts into two or more tests, so less smart people like me have easier time following it? :)

**Comment by tidal** [comment]
@Grunet 
Nice work!
Just a few nit-picks from my side (including my usual ones about temporary variables).
Please, take a look and tell me what you think.

File: src/Contrib/Jaeger/BatchAdapter/BatchAdapter.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+use Jaeger\Thrift\Batch;
+use Thrift\Protocol\TProtocol;
+
+class BatchAdapter implements BatchAdapterInterface
+{
+    private Batch $batchInstance;
+
+    public function __construct(array $vals)
+    {
+        $this->batchInstance = new Batch($vals);
```
[the remark this answers is not in the document]
Comment (Grunet) [reply src/Contrib/Jaeger/BatchAdapter/BatchAdapter.php]: Yup done

File: src/Contrib/Jaeger/BatchAdapter/BatchAdapterFactory.php [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Jaeger\BatchAdapter;
+
+class BatchAdapterFactory implements BatchAdapterFactoryInterface
+{
+    public function createBatchAdapter(array $vals): BatchAdapterInterface
+    {
+        return new BatchAdapter($vals);
```
[the remark this answers is not in the document]
Comment (Grunet) [reply src/Contrib/Jaeger/BatchAdapter/BatchAdapterFactory.php]: Yup done

File: src/Contrib/Jaeger/BatchAdapter/BatchAdapterFactoryInterface.php [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Jaeger\BatchAdapter;
+
+interface BatchAdapterFactoryInterface
+{
+    public function createBatchAdapter(array $vals): BatchAdapterInterface;
```
[the remark this answers is not in the document]
Comment (Grunet) [reply src/Contrib/Jaeger/BatchAdapter/BatchAdapterFactoryInterface.php]: Yeah I did a quick check elsewhere in the repo and I think I saw create*** somewhere which is why I went for this, but I'm guessing there are "create" ones too.

I don't really have a preference, so I switched it to "create"

File: src/Contrib/Jaeger/HttpSender.php (hunk 1 above)
Comment (Grunet) [reply src/Contrib/Jaeger/HttpSender.php]: Yup done

File: src/Contrib/Jaeger/HttpSender.php (hunk 2 above)
Comment (Grunet) [reply src/Contrib/Jaeger/HttpSender.php:98]: Yup done, that looks much better!

File: src/Contrib/Jaeger/TagFactory/TagFactory.php (hunk 3 above)
Comment (Grunet) [reply src/Contrib/Jaeger/TagFactory/TagFactory.php]: Yup done

File: src/SDK/Resource/ResourceInfo.php (hunk 4 above)
Comment (Grunet) [reply src/SDK/Resource/ResourceInfo.php]: Yup done, much cleaner!

File: tests/Unit/Contrib/JaegerHttpSenderTest.php (hunk 5 above)
Comment (Grunet) [reply tests/Unit/Contrib/JaegerHttpSenderTest.php]: Haha I too was a little confused coming back to them...split the assertions across 3 tests, lmk if there's anything else I can do to enhance their understandability

**Comment by Grunet** [comment]
> @Grunet Nice work! Just a few nit-picks from my side (including my usual ones about temporary variables). Please, take a look and tell me what you think.

Yup looks much cleaner now to me. Let me know if I missed anything!

**Review by tidal — APPROVED** [review]

**Pull request merged**